### PR TITLE
Generalize ESR to ESR upgrade action

### PR DESCRIPTION
**Pull request by agarciamontoro**
#### Summary
This PR builds on top of #22448, generalizing the Github Action to accommodate arbitrary initial and final versions, as long as a DB dump of the initial version is provided. I created a new dump for 6.3, which is the same as the one for 5.37, simply upgraded.

Leveraging the now generalized action, the ESR to ESR scripts now tested are:
- 5.37 -> 6.3
- 6.3 -> 7.8
- 5.37 -> 7.8

Each of those scripts have a corresponding cleanup script to unify the dumps of the DB upgraded both from the server and from the script.

While this is in review, I'll work on squashing the main script we wanted to modify, 5.37 -> 7.8 (applying those modifications to 5.37>6.3 and 6.3>7.8 if it's possible).

#### Ticket Link
--

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: .github/workflows/esrupgrade-common.yml [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+          # otherwise, the name of the container gets written to the console, which is weird
+          docker stop mmserver > /dev/null
+      - name: Cleanup DB
+        run : |
+          cd scripts/esrupgrades
+          docker exec -i ghactions_mysql_1 mysql -AN $MYSQL_CONN_ARGS < $CLEANUP_SCRIPT
+      - name: Dump upgraded database
+        run: |
+          # Ignore tables containing meta-information about the migrations (db_{lock,migrations}),
+          # since those will not be present in the upgrade from the script.
```
Comment (lieut-data) [inline .github/workflows/esrupgrade-common.yml:78]: Should we remove this if it's now covered by the `$CLEANUP_SCRIPT`?

File: .github/workflows/esrupgrade.yml [hunk 2]
```diff
@@ -9,144 +9,25 @@ on:
...
-          gzip -d upgraded-dump-server/dump.7.8.server.sql.gz
-          gzip -d upgraded-dump-script/dump.7.8.script.sql.gz
-          diff upgraded-dump-server/dump.7.8.server.sql upgraded-dump-script/dump.7.8.script.sql > diff.7.8.sql
-      - name: Upload diff
-        if: failure() # Upload the diff only if the previous step failed; i.e., if the diff is non-empty
-        uses: actions/upload-artifact@v3
-        with:
-          name: dumps-diff
-          path: diff.7.8.sql
+  esr-upgrade-5_37-7_8:
```
Comment (lieut-data) [inline .github/workflows/esrupgrade.yml:13]: Thinking out loud, but given the overall complexity, I wonder if we should add a 'should-fail' job (and maybe an 'expect-failure' parameter)? The idea being to explicitly test the script itself by providing inputs that are expected to trigger a diff, and just "pass" that. Thoughts?

**Review by lieut-data — APPROVED** [review]
Building on a great foundation!

**Review by isacikgoz — APPROVED** [review]
Perfect, this is an amazing piece of work 💯

File: .github/workflows/esrupgrade.yml (hunk 2 above)
Comment (isacikgoz) [reply .github/workflows/esrupgrade.yml:13]: There are numerous fail scenarios but did you had something specific in your mind? I'm thinking cases we explicitly don't touch. Eg. collation issues etc. Is it something like that?

Comment (lieut-data) [reply .github/workflows/esrupgrade.yml:13]: I'm probably just being paranoid, but given all the moving parts, I was just worried that maybe a regression might result in the tests doing nothing, but they would still pass under current semantics.

**Comment by agarciamontoro** [comment]
How come that merging #22590, that passed all tests, into this branch, that passed all tests, made the tests... fail? Well, I've spent a while trying to understand it, and got the answer now. The diff for all failing tests was just a long list of lines similar to this one:
```sql
> INSERT INTO `Sessions` VALUES ('15gcwczbhf8rmewgqsx8mj84cc','ekt9x9ispbr7zqfkiua6pbzp5o',1677695260662,1680287260662,1677695260662,'1xt14bfunbgkmpddrst7xzx9xa','','system_user',0,0,'{\"os\": \"\", \"csrf\": \"9f5eki548jyaibtsmorgw9eueo\", \"isSaml\": \"false\", \"browser\": \"Unknown/0.0\", \"isMobile\": \"false\", \"is_guest\": \"false\", \"platform\": \"Windows\", \"isOAuthUser\": \"false\"}');
```
The migration through the script kept these lines intact, while the server one removed all of them. It makes sense that the server got rid of them, since the server has to delete all sessions that are expired. But why didn't it delete them in the CI that executed a couple of weeks ago, while it did remove them in this one? The reason is simple: the `ExpiresAt` value in all those lines is similar to 1680287260662, which happens to be Friday, March 31, 2023 18:27:40.662 UTC.

The latest CI for #22590 ran on Friday, March 31, 2023... at 16:40 UTC! Less than a couple of hours before the sessions expired! That's why they passed then, but not now :) This was a fun investigation.

Anyway, I'll need to fix this additional quirk, and I think the best way to do so is to preprocess the script modifying the `ExpiresAt` column. We could alternatively remove all rows in the Sessions table in the cleanup phase, but that could hide some modifications to this table, so I'll go with the new preprocess phase.

File: .github/workflows/esrupgrade-common.yml (hunk 1 above)
Comment (agarciamontoro) [reply .github/workflows/esrupgrade-common.yml]: Thanks! Done in https://github.com/mattermost/mattermost-server/pull/22573/commits/7552c5dd63d64b06e9cfd834edcf479490ead2de.

**Comment by agarciamontoro** [comment]
Huh, maybe I shouldn't have declared victory so soon. I'll keep investigating :eyes:

**Comment by agarciamontoro** [comment]
Cool, just re-running CI made it pass, so it's indeed a flaky test. I'll try to fix it in the preprocess phase.

**Comment by agarciamontoro** [comment]
Phew, it passed. At the end, I had to add three modifications:
1. Preprocessing: Update all rows in Sessions so that they never expire (setting `ExpiresAt` to 0), forcing the server to keep them. See https://github.com/mattermost/mattermost-server/pull/22573/commits/2b1abf267295d7a87566bc484cd3fc3d7e1e0c7e and https://github.com/mattermost/mattermost-server/pull/22573/commits/3913e9c656130b97b7083f969bcf173196b7ebe8
2. Preprocessing: Add a `system-bot` user if it does not exist, forcing the server to use it instead of creating a new one that may not be in the dump. See https://github.com/mattermost/mattermost-server/pull/22573/commits/7a6a16f096e49aebca9659fd9296d1c4ab8754bc
3. Postprocessing: Delete all rows in the ProductNoticeViewState table, which is populated with info from an external source (https://github.com/mattermost/notices) that we cannot control See https://github.com/mattermost/mattermost-server/pull/22573/commits/906dbcfe365f96e049598cd1950ab874f56ddbe5 (and a typo fix in https://github.com/mattermost/mattermost-server/pull/22573/commits/f5c682e5be5c058ed3571164c4ad49e452fea384)

@lieut-data, @isacikgoz: sorry for the re-request of review, but I'd like you to take a look to these commits before merging.

About [the failing step in CI](https://github.com/mattermost/mattermost-server/actions/runs/4687930230/jobs/8307839260?pr=22573): I'll deal with it when I merge this branch into #22448 (I'm hoping the error magically goes away when running against the monorepo).

**Review by isacikgoz — APPROVED** [review]
This looks good, wondering if we could use `mattermost db migrate` command to not involve application at all. WDYT @agarciamontoro?

**Comment by isacikgoz** [comment]
> @isacikgoz: Ah, that's a good idea, but I'm a bit concerned the logic in that command ends up deviating from what the server does, even if slightly. I took a look and it simply calls `sqlstore.New`, so it should be the same, but what if it changes in the future? 🤔 Am I being too paranoid, perhaps?

Yeah I know, we need to be a bit paranoid here. My thinking was just to avoid unnecessary application level changes like the session issue and system bot thing. Maybe we can add this to backlog for future imporovements.

**Review by lieut-data — APPROVED** [review]

**Pull request merged**